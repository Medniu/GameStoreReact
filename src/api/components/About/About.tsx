import { ReactElement } from "react";
import Container from "../Container/Container";

function About(): ReactElement {
  return (
    <div>
      <Container>
        <h1>Some info about site</h1>
      </Container>
    </div>
  );
}

export default About;
