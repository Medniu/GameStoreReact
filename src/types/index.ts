interface TopGames {
  id: number;
  name: string;
  price: number;
  image: string;
  description: string;
  category: string;
  rating: number;
}

interface Params {
  category: string;
}

type Item = {
  id: number;
  name: string;
  price: number;
  description: string;
  category: string;
  rating: number;
  image: string;
};

type User = {
  photo: string;
  login: string;
  address: string;
  phoneNumber: string;
};

type AuthState = {
  user: User | null | undefined;
};

type AllState = {
  auth: AuthState;
};

type Action = {
  type: string;
  payload: User | null | undefined;
};

export { User, AuthState, AllState, TopGames, Params, Item, Action };
