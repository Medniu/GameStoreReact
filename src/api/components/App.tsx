import { Route, Switch, Redirect, withRouter, RouteComponentProps } from "react-router-dom";
import React, { ReactElement } from "react";
import Home from "./Home/Home";
import Products from "./Products/Products";
import About from "./About/About";
import Alert from "./Alert/Alert";
import "../../styles/global.css";
import ProfilePage from "./ProfilePage/ProfilePage";

type State = {
  hasError: boolean;
};

type AppProps = RouteComponentProps;

class App extends React.Component<AppProps, State> {
  constructor(props: AppProps) {
    super(props);
    this.state = { hasError: false };
  }

  componentDidCatch() {
    this.setState({ hasError: true });
  }

  handleAlertState = (hasError: boolean) => {
    this.setState({ hasError });
    this.props.history.push("/");
  };

  render(): ReactElement {
    return (
      <>
        <Switch>
          <Route exact path="/" component={() => <Home />} />
          <Route path="/products/:category" component={() => <Products />} />
          <Route path="/About" component={() => <About />} />
          <Route path="/Profile" component={() => <ProfilePage />} />
          <Route render={() => <Redirect to={{ pathname: "/" }} />} />
        </Switch>
        {this.state.hasError && (
          <Alert
            title="Error Title"
            hasError={this.state.hasError}
            errorMessage="Some Error"
            setActive={this.handleAlertState}
          />
        )}
      </>
    );
  }
}
export default withRouter(App);
