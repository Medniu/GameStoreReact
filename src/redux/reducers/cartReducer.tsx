import { CartState, CartAction, CartItem } from "../../types";

const INITIAL_STATE: CartState = {
  items: [],
  total: 0,
};

export default (state = INITIAL_STATE, action: CartAction): CartState => {
  switch (action.type) {
    case "ADD_TO_CART": {
      const existedItem = state.items.find((item) => action.payload?.id === item.id);
      if (existedItem && action.payload) {
        existedItem.quantity += 1;
        return {
          ...state,
          total: state.total + action.payload?.price,
        };
      }
      if (action.payload) {
        const newTotal = state.total + action.payload?.price;
        const newCartItem: CartItem = {
          id: action.payload?.id,
          name: action.payload?.name,
          price: action.payload?.price,
          category: action.payload?.category,
          quantity: 1,
          isSelected: false,
        };
        return { ...state, items: [...state.items, newCartItem], total: newTotal };
      }
      return state;
    }

    case "BUY_ITEMS": {
      const newList: Array<CartItem> = [];

      return { ...state, items: newList, total: 0 };
    }
    case "REMOVE_ITEMS": {
      const newList = state.items.filter((item) => item.isSelected === false);
      if (newList) {
        let newTotal = 0;
        newList.forEach((item) => {
          newTotal += item.price * item.quantity;
        });
        return { ...state, items: newList, total: newTotal };
      }
      return state;
    }
    case "ADD_QUANTITY": {
      const addedItem = state.items.find((item) => action.payload?.id === item.id);
      if (addedItem) {
        addedItem.quantity += 1;
        const newTotal = state.total + addedItem.price;
        return {
          ...state,
          total: newTotal,
        };
      }
      return state;
    }
    case "REMOVE_QUANTITY": {
      const removedItem = state.items.find((item) => action.payload?.id === item.id);
      if (removedItem?.quantity === 1) {
        const newList = state.items.filter((item) => action.payload?.id !== item.id);
        let newTotal = 0;
        newList.forEach((item) => {
          newTotal += item.price * item.quantity;
        });
        return { ...state, items: newList, total: newTotal };
      }
      if (removedItem && removedItem.quantity >= 1) {
        removedItem.quantity -= 1;
        const newTotal = state.total - removedItem.price;
        return {
          ...state,
          total: newTotal,
        };
      }
      return state;
    }
    case "SELECT_ITEM": {
      const selectedItem = state.items.find((item) => action.payload?.id === item.id);
      if (selectedItem) {
        selectedItem.isSelected = !selectedItem.isSelected;
        return { ...state, items: [...state.items] };
      }
      return state;
    }
    default:
      return state;
  }
};
