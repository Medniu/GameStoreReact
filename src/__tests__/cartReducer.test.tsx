import React from "react";
import ReactDOM from "react-dom";
import { CartItem, CartState, Item } from "@/types";
import App from "../api/components/App";
import About from "../api/components/About/About";
import cartReducer from "../redux/reducers/cartReducer";

describe("cartReducer tests", () => {
  it("It should add new item to cart", () => {
    const state: CartState = {
      items: [],
      total: 0,
    };
    const newCartItem: Item = {
      id: 1,
      name: "Witcher",
      price: 80,
      category: "PC",
      description: "das",
      rating: 5,
      image: "dasdas",
      genre: "shooter",
      age: "18",
    };
    const newState = cartReducer(state, {
      type: "ADD_TO_CART",
      payload: newCartItem,
    });

    expect(newState.items.length).toBe(1);
  });
  it("It should buy items and remove them from cart", () => {
    const state: CartState = {
      items: [
        {
          id: 0,
          name: "Witcher",
          price: 80,
          category: "PC",
          quantity: 2,
          isSelected: true,
        },
        {
          id: 1,
          name: "Witcher 2",
          price: 30,
          category: "PC",
          quantity: 1,
          isSelected: true,
        },
        {
          id: 2,
          name: "Witcher",
          price: 50,
          category: "PC",
          quantity: 2,
          isSelected: true,
        },
      ],
      total: 290,
    };
    const newState = cartReducer(state, {
      type: "BUY_ITEMS",
      payload: null,
    });

    expect(newState.items.length).toBe(0);
    expect(newState.total).toBe(0);
  });
  it("It should change isSelected to true", () => {
    const state: CartState = {
      items: [
        {
          id: 0,
          name: "Witcher",
          price: 80,
          category: "PC",
          quantity: 2,
          isSelected: false,
        },
        {
          id: 1,
          name: "Witcher 2",
          price: 30,
          category: "PC",
          quantity: 1,
          isSelected: false,
        },
      ],
      total: 190,
    };
    const newSelectedCartItem: Item = {
      id: 1,
      name: "Witcher",
      price: 80,
      category: "PC",
      description: "das",
      rating: 5,
      image: "dasdas",
      genre: "shooter",
      age: "18",
    };
    const newState = cartReducer(state, {
      type: "SELECT_ITEM",
      payload: newSelectedCartItem,
    });
    const updatedItem = newState.items.find((item) => item.id === newSelectedCartItem.id);
    expect(updatedItem?.isSelected).toBe(true);
    expect(state).toMatchSnapshot();
    expect(newState).toMatchSnapshot();
    expect(updatedItem).toMatchSnapshot();
  });
});
