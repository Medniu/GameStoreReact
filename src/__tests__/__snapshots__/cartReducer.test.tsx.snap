// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cartReducer tests It should change isSelected to true 1`] = `
Object {
  "items": Array [
    Object {
      "category": "PC",
      "id": 0,
      "isSelected": false,
      "name": "Witcher",
      "price": 80,
      "quantity": 2,
    },
    Object {
      "category": "PC",
      "id": 1,
      "isSelected": true,
      "name": "Witcher 2",
      "price": 30,
      "quantity": 1,
    },
  ],
  "total": 190,
}
`;

exports[`cartReducer tests It should change isSelected to true 2`] = `
Object {
  "items": Array [
    Object {
      "category": "PC",
      "id": 0,
      "isSelected": false,
      "name": "Witcher",
      "price": 80,
      "quantity": 2,
    },
    Object {
      "category": "PC",
      "id": 1,
      "isSelected": true,
      "name": "Witcher 2",
      "price": 30,
      "quantity": 1,
    },
  ],
  "total": 190,
}
`;

exports[`cartReducer tests It should change isSelected to true 3`] = `
Object {
  "category": "PC",
  "id": 1,
  "isSelected": true,
  "name": "Witcher 2",
  "price": 30,
  "quantity": 1,
}
`;
